lib/tiny-jsonrpc/server.js

```diff
@@ -260,25 +285,66 @@ Server.prototype._respond = function (request) {
       !util.isObject(request.params)
     )
   ) {
-    return this.makeError(util.isUndefined(request.id) ?
-      null : request.id,
-      this.errors.INVALID_REQUEST);
+    return callback(
+      null,
+      this.makeError(util.isUndefined(request.id) ?
+        null : request.id,
+        this.errors.INVALID_REQUEST
+      )
+    );
   }
 
   // coolbro. handle it.
   var method = this._methods[request.method];
   if (!method) {
-    return this.makeError(request.id, this.errors.METHOD_NOT_FOUND);
+    return callback(
+      null,
+      this.makeError(request.id, this.errors.METHOD_NOT_FOUND)
+    );
   }
 
+  var that = this;
+  var methodCallback = function (error, result) {
+    var code;
+
+    if (error) {
+      code = util.isNumber(error.code) ?
+        error.code :
+        that.errors.INTERNAL_ERROR;
+
+      return callback(
+        null,
+        that.makeError(
+          request.id,
+          error.message || error,
+          error.code,
+          error.data
+        )
+      );
+    }
+
+    return callback(
+      null,
+      that.makeResponse(request.id, result)
+    );
+  };
+  var isAsync = false;
+
+  var context = {
+    async: function () {
+      isAsync = true;
+      return methodCallback;
+    }
+  };
+  var result;
+
   try {
-    return this.makeResponse(request.id,
-      method.fn.apply(null, marshal(request.params, method))
+    methodCallback(
+      null,
+      method.fn.apply(context, marshal(request.params, method))
```

Comment:
When the provided method is async, it looks like we'll call `callback` twice. The first time will happen here, and the second time happens when the provided method does `this.async()(null, result)`.